File: py/bluefintecsuserbackoffice_sdk.py
# BluefinTecsUserBackoffice SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import BluefinTecsUserBackofficeUtility
from core.spec import BluefinTecsUserBackofficeSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import BluefinTecsUserBackofficeBaseFeature
from features import _make_feature


class BluefinTecsUserBackofficeSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = BluefinTecsUserBackofficeUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return BluefinTecsUserBackofficeUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = BluefinTecsUserBackofficeSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def OutputActivateDigitalModule(self, data=None) -> "OutputActivateDigitalModuleEntity":
        """Entity factory: client.OutputActivateDigitalModule().list() / client.OutputActivateDigitalModule().load({"id": ...})."""
        from entity.output_activate_digital_module_entity import OutputActivateDigitalModuleEntity
        return OutputActivateDigitalModuleEntity(self, data)


    def OutputActivatePortalModule(self, data=None) -> "OutputActivatePortalModuleEntity":
        """Entity factory: client.OutputActivatePortalModule().list() / client.OutputActivatePortalModule().load({"id": ...})."""
        from entity.output_activate_portal_module_entity import OutputActivatePortalModuleEntity
        return OutputActivatePortalModuleEntity(self, data)


    def OutputActivateStoreModule(self, data=None) -> "OutputActivateStoreModuleEntity":
        """Entity factory: client.OutputActivateStoreModule().list() / client.OutputActivateStoreModule().load({"id": ...})."""
        from entity.output_activate_store_module_entity import OutputActivateStoreModuleEntity
        return OutputActivateStoreModuleEntity(self, data)


    def OutputActivateUser(self, data=None) -> "OutputActivateUserEntity":
        """Entity factory: client.OutputActivateUser().list() / client.OutputActivateUser().load({"id": ...})."""
        from entity.output_activate_user_entity import OutputActivateUserEntity
        return OutputActivateUserEntity(self, data)


    def OutputAssignRole(self, data=None) -> "OutputAssignRoleEntity":
        """Entity factory: client.OutputAssignRole().list() / client.OutputAssignRole().load({"id": ...})."""
        from entity.output_assign_role_entity import OutputAssignRoleEntity
        return OutputAssignRoleEntity(self, data)


    def OutputChangeLogo(self, data=None) -> "OutputChangeLogoEntity":
        """Entity factory: client.OutputChangeLogo().list() / client.OutputChangeLogo().load({"id": ...})."""
        from entity.output_change_logo_entity import OutputChangeLogoEntity
        return OutputChangeLogoEntity(self, data)


    def OutputCreateMandator(self, data=None) -> "OutputCreateMandatorEntity":
        """Entity factory: client.OutputCreateMandator().list() / client.OutputCreateMandator().load({"id": ...})."""
        from entity.output_create_mandator_entity import OutputCreateMandatorEntity
        return OutputCreateMandatorEntity(self, data)


    def OutputCreateServiceUser(self, data=None) -> "OutputCreateServiceUserEntity":
        """Entity factory: client.OutputCreateServiceUser().list() / client.OutputCreateServiceUser().load({"id": ...})."""
        from entity.output_create_service_user_entity import OutputCreateServiceUserEntity
        return OutputCreateServiceUserEntity(self, data)


    def OutputDeactivateUser(self, data=None) -> "OutputDeactivateUserEntity":
        """Entity factory: client.OutputDeactivateUser().list() / client.OutputDeactivateUser().load({"id": ...})."""
        from entity.output_deactivate_user_entity import OutputDeactivateUserEntity
        return OutputDeactivateUserEntity(self, data)


    def OutputGetKycDocument(self, data=None) -> "OutputGetKycDocumentEntity":
        """Entity factory: client.OutputGetKycDocument().list() / client.OutputGetKycDocument().load({"id": ...})."""
        from entity.output_get_kyc_document_entity import OutputGetKycDocumentEntity
        return OutputGetKycDocumentEntity(self, data)


    def OutputGetLogo(self, data=None) -> "OutputGetLogoEntity":
        """Entity factory: client.OutputGetLogo().list() / client.OutputGetLogo().load({"id": ...})."""
        from entity.output_get_logo_entity import OutputGetLogoEntity
        return OutputGetLogoEntity(self, data)


    def OutputListOfAvailableRole(self, data=None) -> "OutputListOfAvailableRoleEntity":
        """Entity factory: client.OutputListOfAvailableRole().list() / client.OutputListOfAvailableRole().load({"id": ...})."""
        from entity.output_list_of_available_role_entity import OutputListOfAvailableRoleEntity
        return OutputListOfAvailableRoleEntity(self, data)


    def OutputListOfMandator(self, data=None) -> "OutputListOfMandatorEntity":
        """Entity factory: client.OutputListOfMandator().list() / client.OutputListOfMandator().load({"id": ...})."""
        from entity.output_list_of_mandator_entity import OutputListOfMandatorEntity
        return OutputListOfMandatorEntity(self, data)


    def OutputListOfModule(self, data=None) -> "OutputListOfModuleEntity":
        """Entity factory: client.OutputListOfModule().list() / client.OutputListOfModule().load({"id": ...})."""
        from entity.output_list_of_module_entity import OutputListOfModuleEntity
        return OutputListOfModuleEntity(self, data)


    def OutputListOfRoleGroup(self, data=None) -> "OutputListOfRoleGroupEntity":
        """Entity factory: client.OutputListOfRoleGroup().list() / client.OutputListOfRoleGroup().load({"id": ...})."""
        from entity.output_list_of_role_group_entity import OutputListOfRoleGroupEntity
        return OutputListOfRoleGroupEntity(self, data)


    def OutputListOfTransactionsHistory(self, data=None) -> "OutputListOfTransactionsHistoryEntity":
        """Entity factory: client.OutputListOfTransactionsHistory().list() / client.OutputListOfTransactionsHistory().load({"id": ...})."""
        from entity.output_list_of_transactions_history_entity import OutputListOfTransactionsHistoryEntity
        return OutputListOfTransactionsHistoryEntity(self, data)


    def OutputListOfUser(self, data=None) -> "OutputListOfUserEntity":
        """Entity factory: client.OutputListOfUser().list() / client.OutputListOfUser().load({"id": ...})."""
        from entity.output_list_of_user_entity import OutputListOfUserEntity
        return OutputListOfUserEntity(self, data)


    def OutputProvideCredential(self, data=None) -> "OutputProvideCredentialEntity":
        """Entity factory: client.OutputProvideCredential().list() / client.OutputProvideCredential().load({"id": ...})."""
        from entity.output_provide_credential_entity import OutputProvideCredentialEntity
        return OutputProvideCredentialEntity(self, data)


    def OutputRegisterUser(self, data=None) -> "OutputRegisterUserEntity":
        """Entity factory: client.OutputRegisterUser().list() / client.OutputRegisterUser().load({"id": ...})."""
        from entity.output_register_user_entity import OutputRegisterUserEntity
        return OutputRegisterUserEntity(self, data)


    def OutputRemoveRole(self, data=None) -> "OutputRemoveRoleEntity":
        """Entity factory: client.OutputRemoveRole().list() / client.OutputRemoveRole().load({"id": ...})."""
        from entity.output_remove_role_entity import OutputRemoveRoleEntity
        return OutputRemoveRoleEntity(self, data)


    def OutputResendLink(self, data=None) -> "OutputResendLinkEntity":
        """Entity factory: client.OutputResendLink().list() / client.OutputResendLink().load({"id": ...})."""
        from entity.output_resend_link_entity import OutputResendLinkEntity
        return OutputResendLinkEntity(self, data)


    def OutputResetPassword(self, data=None) -> "OutputResetPasswordEntity":
        """Entity factory: client.OutputResetPassword().list() / client.OutputResetPassword().load({"id": ...})."""
        from entity.output_reset_password_entity import OutputResetPasswordEntity
        return OutputResetPasswordEntity(self, data)


    def OutputUpdateConsumer(self, data=None) -> "OutputUpdateConsumerEntity":
        """Entity factory: client.OutputUpdateConsumer().list() / client.OutputUpdateConsumer().load({"id": ...})."""
        from entity.output_update_consumer_entity import OutputUpdateConsumerEntity
        return OutputUpdateConsumerEntity(self, data)


    def OutputUpdateProfile(self, data=None) -> "OutputUpdateProfileEntity":
        """Entity factory: client.OutputUpdateProfile().list() / client.OutputUpdateProfile().load({"id": ...})."""
        from entity.output_update_profile_entity import OutputUpdateProfileEntity
        return OutputUpdateProfileEntity(self, data)


    def Version(self, data=None) -> "VersionEntity":
        """Entity factory: client.Version().list() / client.Version().load({"id": ...})."""
        from entity.version_entity import VersionEntity
        return VersionEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "BluefinTecsUserBackofficeSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.output_activate_digital_module_entity import OutputActivateDigitalModuleEntity
    from entity.output_activate_portal_module_entity import OutputActivatePortalModuleEntity
    from entity.output_activate_store_module_entity import OutputActivateStoreModuleEntity
    from entity.output_activate_user_entity import OutputActivateUserEntity
    from entity.output_assign_role_entity import OutputAssignRoleEntity
    from entity.output_change_logo_entity import OutputChangeLogoEntity
    from entity.output_create_mandator_entity import OutputCreateMandatorEntity
    from entity.output_create_service_user_entity import OutputCreateServiceUserEntity
    from entity.output_deactivate_user_entity import OutputDeactivateUserEntity
    from entity.output_get_kyc_document_entity import OutputGetKycDocumentEntity
    from entity.output_get_logo_entity import OutputGetLogoEntity
    from entity.output_list_of_available_role_entity import OutputListOfAvailableRoleEntity
    from entity.output_list_of_mandator_entity import OutputListOfMandatorEntity
    from entity.output_list_of_module_entity import OutputListOfModuleEntity
    from entity.output_list_of_role_group_entity import OutputListOfRoleGroupEntity
    from entity.output_list_of_transactions_history_entity import OutputListOfTransactionsHistoryEntity
    from entity.output_list_of_user_entity import OutputListOfUserEntity
    from entity.output_provide_credential_entity import OutputProvideCredentialEntity
    from entity.output_register_user_entity import OutputRegisterUserEntity
    from entity.output_remove_role_entity import OutputRemoveRoleEntity
    from entity.output_resend_link_entity import OutputResendLinkEntity
    from entity.output_reset_password_entity import OutputResetPasswordEntity
    from entity.output_update_consumer_entity import OutputUpdateConsumerEntity
    from entity.output_update_profile_entity import OutputUpdateProfileEntity
    from entity.version_entity import VersionEntity

File: rb/BluefinTecsUserBackoffice_sdk.rb
# BluefinTecsUserBackoffice SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'BluefinTecsUserBackoffice_types'


class BluefinTecsUserBackofficeSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = BluefinTecsUserBackofficeUtility.new
    @_utility = utility

    config = BluefinTecsUserBackofficeConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = BluefinTecsUserBackofficeHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = BluefinTecsUserBackofficeHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, BluefinTecsUserBackofficeFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    BluefinTecsUserBackofficeUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = BluefinTecsUserBackofficeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = BluefinTecsUserBackofficeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = BluefinTecsUserBackofficeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = BluefinTecsUserBackofficeSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue BluefinTecsUserBackofficeError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = BluefinTecsUserBackofficeHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = BluefinTecsUserBackofficeHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.OutputActivateDigitalModule.list / client.OutputActivateDigitalModule.load({ "id" => ... })
  def OutputActivateDigitalModule(data = nil)
    require_relative 'entity/output_activate_digital_module_entity'
    OutputActivateDigitalModuleEntity.new(self, data)
  end


  # Canonical facade: client.OutputActivatePortalModule.list / client.OutputActivatePortalModule.load({ "id" => ... })
  def OutputActivatePortalModule(data = nil)
    require_relative 'entity/output_activate_portal_module_entity'
    OutputActivatePortalModuleEntity.new(self, data)
  end


  # Canonical facade: client.OutputActivateStoreModule.list / client.OutputActivateStoreModule.load({ "id" => ... })
  def OutputActivateStoreModule(data = nil)
    require_relative 'entity/output_activate_store_module_entity'
    OutputActivateStoreModuleEntity.new(self, data)
  end


  # Canonical facade: client.OutputActivateUser.list / client.OutputActivateUser.load({ "id" => ... })
  def OutputActivateUser(data = nil)
    require_relative 'entity/output_activate_user_entity'
    OutputActivateUserEntity.new(self, data)
  end


  # Canonical facade: client.OutputAssignRole.list / client.OutputAssignRole.load({ "id" => ... })
  def OutputAssignRole(data = nil)
    require_relative 'entity/output_assign_role_entity'
    OutputAssignRoleEntity.new(self, data)
  end


  # Canonical facade: client.OutputChangeLogo.list / client.OutputChangeLogo.load({ "id" => ... })
  def OutputChangeLogo(data = nil)
    require_relative 'entity/output_change_logo_entity'
    OutputChangeLogoEntity.new(self, data)
  end


  # Canonical facade: client.OutputCreateMandator.list / client.OutputCreateMandator.load({ "id" => ... })
  def OutputCreateMandator(data = nil)
    require_relative 'entity/output_create_mandator_entity'
    OutputCreateMandatorEntity.new(self, data)
  end


  # Canonical facade: client.OutputCreateServiceUser.list / client.OutputCreateServiceUser.load({ "id" => ... })
  def OutputCreateServiceUser(data = nil)
    require_relative 'entity/output_create_service_user_entity'
    OutputCreateServiceUserEntity.new(self, data)
  end


  # Canonical facade: client.OutputDeactivateUser.list / client.OutputDeactivateUser.load({ "id" => ... })
  def OutputDeactivateUser(data = nil)
    require_relative 'entity/output_deactivate_user_entity'
    OutputDeactivateUserEntity.new(self, data)
  end


  # Canonical facade: client.OutputGetKycDocument.list / client.OutputGetKycDocument.load({ "id" => ... })
  def OutputGetKycDocument(data = nil)
    require_relative 'entity/output_get_kyc_document_entity'
    OutputGetKycDocumentEntity.new(self, data)
  end


  # Canonical facade: client.OutputGetLogo.list / client.OutputGetLogo.load({ "id" => ... })
  def OutputGetLogo(data = nil)
    require_relative 'entity/output_get_logo_entity'
    OutputGetLogoEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfAvailableRole.list / client.OutputListOfAvailableRole.load({ "id" => ... })
  def OutputListOfAvailableRole(data = nil)
    require_relative 'entity/output_list_of_available_role_entity'
    OutputListOfAvailableRoleEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfMandator.list / client.OutputListOfMandator.load({ "id" => ... })
  def OutputListOfMandator(data = nil)
    require_relative 'entity/output_list_of_mandator_entity'
    OutputListOfMandatorEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfModule.list / client.OutputListOfModule.load({ "id" => ... })
  def OutputListOfModule(data = nil)
    require_relative 'entity/output_list_of_module_entity'
    OutputListOfModuleEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfRoleGroup.list / client.OutputListOfRoleGroup.load({ "id" => ... })
  def OutputListOfRoleGroup(data = nil)
    require_relative 'entity/output_list_of_role_group_entity'
    OutputListOfRoleGroupEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfTransactionsHistory.list / client.OutputListOfTransactionsHistory.load({ "id" => ... })
  def OutputListOfTransactionsHistory(data = nil)
    require_relative 'entity/output_list_of_transactions_history_entity'
    OutputListOfTransactionsHistoryEntity.new(self, data)
  end


  # Canonical facade: client.OutputListOfUser.list / client.OutputListOfUser.load({ "id" => ... })
  def OutputListOfUser(data = nil)
    require_relative 'entity/output_list_of_user_entity'
    OutputListOfUserEntity.new(self, data)
  end


  # Canonical facade: client.OutputProvideCredential.list / client.OutputProvideCredential.load({ "id" => ... })
  def OutputProvideCredential(data = nil)
    require_relative 'entity/output_provide_credential_entity'
    OutputProvideCredentialEntity.new(self, data)
  end


  # Canonical facade: client.OutputRegisterUser.list / client.OutputRegisterUser.load({ "id" => ... })
  def OutputRegisterUser(data = nil)
    require_relative 'entity/output_register_user_entity'
    OutputRegisterUserEntity.new(self, data)
  end


  # Canonical facade: client.OutputRemoveRole.list / client.OutputRemoveRole.load({ "id" => ... })
  def OutputRemoveRole(data = nil)
    require_relative 'entity/output_remove_role_entity'
    OutputRemoveRoleEntity.new(self, data)
  end


  # Canonical facade: client.OutputResendLink.list / client.OutputResendLink.load({ "id" => ... })
  def OutputResendLink(data = nil)
    require_relative 'entity/output_resend_link_entity'
    OutputResendLinkEntity.new(self, data)
  end


  # Canonical facade: client.OutputResetPassword.list / client.OutputResetPassword.load({ "id" => ... })
  def OutputResetPassword(data = nil)
    require_relative 'entity/output_reset_password_entity'
    OutputResetPasswordEntity.new(self, data)
  end


  # Canonical facade: client.OutputUpdateConsumer.list / client.OutputUpdateConsumer.load({ "id" => ... })
  def OutputUpdateConsumer(data = nil)
    require_relative 'entity/output_update_consumer_entity'
    OutputUpdateConsumerEntity.new(self, data)
  end


  # Canonical facade: client.OutputUpdateProfile.list / client.OutputUpdateProfile.load({ "id" => ... })
  def OutputUpdateProfile(data = nil)
    require_relative 'entity/output_update_profile_entity'
    OutputUpdateProfileEntity.new(self, data)
  end


  # Canonical facade: client.Version.list / client.Version.load({ "id" => ... })
  def Version(data = nil)
    require_relative 'entity/version_entity'
    VersionEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = BluefinTecsUserBackofficeSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
